org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java

```diff
@@ -89,10 +97,71 @@ public void apply(final Multimap<String, String> record, final List<String> para
     remove_field {
         public void apply(final Multimap<String, String> record, final List<String> params,
                 final Map<String, String> options) {
-            record.removeAll(params.get(0));
+            params.forEach(p -> {
+                record.removeAll(p);
+            });
+        }
+    },
+    format {
+        public void apply(final Multimap<String, String> record, final List<String> params,
+                final Map<String, String> options) {
+            final Collection<String> oldVals = record.get(params.get(0));
+            final String newVal = String.format(params.get(1), oldVals.toArray(new Object[] {}));
+            record.replaceValues(params.get(0), Arrays.asList(newVal));
         }
     },
+    parse_text {
+        public void apply(final Multimap<String, String> record, final List<String> params,
+                final Map<String, String> options) {
+            record.get(params.get(0)).forEach(v -> {
+                final Pattern p = Pattern.compile(params.get(1));
+                final Matcher m = p.matcher(v);
+                if (m.matches()) {
+                    record.removeAll(params.get(0));
+                    final Map<String, Integer> namedGroups = getNamedGroups(p);
+                    if (!namedGroups.isEmpty()) {
+                        namedGroups.keySet().forEach(k -> {
+                            record.put(params.get(0) + "." + k, m.group(namedGroups.get(k)));
+                        });
+                    }
+                    else {
+                        for (int i = 1; i <= m.groupCount(); i = i + 1) {
+                            record.put(params.get(0), m.group(i));
+                        }
+                    }
+                }
+            });
+        }
 
+        @SuppressWarnings("unchecked")
+        private Map<String, Integer> getNamedGroups(final Pattern regex) {
+            try {
+                // Not available as API, see https://stackoverflow.com/a/15596145/18154:
+                final Method namedGroupsMethod = Pattern.class.getDeclaredMethod("namedGroups");
+                namedGroupsMethod.setAccessible(true);
```

Comment:
This doesn't seem to work when consumed as an external dependency (Maven package or Gradle source dependency; internal tests work fine):

```
java.lang.reflect.InaccessibleObjectException: Unable to make java.util.Map java.util.regex.Pattern.namedGroups() accessible: module java.base does not "opens java.util.regex" to unnamed module @30ee2816
    at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354) ~[?:?]
    at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297) ~[?:?]
    at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:199) ~[?:?]
    at java.base/java.lang.reflect.Method.setAccessible(Method.java:193) ~[?:?]
    at org.metafacture.metafix.FixMethod$11.getNamedGroups(FixMethod.java:168) ~[metafacture-fix-0.2.0-SNAPSHOT.jar:?]
    at org.metafacture.metafix.FixMethod$11.lambda$apply$1(FixMethod.java:146) ~[metafacture-fix-0.2.0-SNAPSHOT.jar:?]
    at java.base/java.util.ArrayList.forEach(ArrayList.java:1511) ~[?:?]
    at org.metafacture.metafix.FixMethod$11.apply(FixMethod.java:141) ~[metafacture-fix-0.2.0-SNAPSHOT.jar:?]
    at org.metafacture.metafix.RecordTransformer.processFunction(RecordTransformer.java:170) ~[metafacture-fix-0.2.0-SNAPSHOT.jar:?]
    at org.metafacture.metafix.RecordTransformer.processSubexpressions(RecordTransformer.java:89) ~[metafacture-fix-0.2.0-SNAPSHOT.jar:?]
    at org.metafacture.metafix.RecordTransformer.transform(RecordTransformer.java:64) ~[metafacture-fix-0.2.0-SNAPSHOT.jar:?]
    at org.metafacture.metafix.Metafix.endRecord(Metafix.java:137) ~[metafacture-fix-0.2.0-SNAPSHOT.jar:?]
    at org.metafacture.flowcontrol.StreamBuffer.replay(StreamBuffer.java:64) ~[metafacture-flowcontrol-5.2.0.jar:?]
    at org.metafacture.metamorph.Filter.dispatch(Filter.java:84) ~[metamorph-5.2.0.jar:?]
    at org.metafacture.metamorph.Filter.endRecord(Filter.java:99) ~[metamorph-5.2.0.jar:?]
[...]
```

Can you confirm? If so, we'd have to find a different solution.

Reply:
It <a href="https://test.metafacture.org/playground/?data=1%7Bdate%3A+2015-03-07%7D&flux=as-lines%0A%7Cdecode-formeta%0A%7Cfix%0A%7Cencode-json&fix=parse_text%28date%2C+%27%28%3F%3Cyear%3E\\d%7B4%7D%29-%28%3F%3Cmonth%3E\\d%7B2%7D%29-%28%3F%3Cday%3E\\d%7B2%7D%29%27%29&morph=%3C%3Fxml+version%3D%221.0%22+encoding%3D%22UTF-8%22%3F%3E%0A%3Cmetamorph+xmlns%3D%22http%3A//www.culturegraph.org/metamorph%22+xmlns%3Axsi%3D%22http%3A//www.w3.org/2001/XMLSchema-instance%22%0A%09version%3D%221%22%3E%0A%09%3Crules%3E%0A%09%09%3Cdata+source%3D%22_id%22+name%3D%22id%22/%3E%0A%09%09%3Cdata+source%3D%22a%22+name%3D%22title%22/%3E%0A%09%09%3Ccombine+value%3D%22%24%7Bfirst%7D+%24%7Blast%7D+aus+%24%7Bplace%7D%22+name%3D%22author%22%3E%0A%09%09%09%3Cdata+source%3D%22b.v%22+name%3D%22first%22+/%3E%0A%09%09%09%3Cdata+source%3D%22b.n%22+name%3D%22last%22+/%3E%0A%09%09%09%3Cdata+source%3D%22c%22+name%3D%22place%22+/%3E%0A%09%09%3C/combine%3E%0A%09%3C/rules%3E%0A%3C/metamorph%3E%0A&active-editor=fix">works in the playground</a>, which consumes metafacture-fix via the local Maven repo, but that's probably because the playground also runs on Java 8, so we will have to find a solution that works with the Java module system.

Reply:
> but that's probably because the playground also runs on Java 8

Yeah, sorry, should have mentioned that. Limetrans is on 17, which puts the onus on me, I guess... :frowning_face:

Reply:
Would you be open to adding a simple guard for now, so we can at least use `parse_text` with numbered groups?

```java
final Map<String, Integer> namedGroups = params.get(1).contains("(?<") ? getNamedGroups(p) : Collections.emptyMap();
```

Then we can postpone the search for a compatible solution.

Reply:
Yes, good idea!

Reply:
I _might_ have an idea for a solution. Stay tuned.

Reply:
See 5395e90 for a possible alternative. Do you think the requirements are reasonable?

Reply:
> 1. Named groups are not escaped/quoted

Not sure what you mean here. Do you mean it won't work to match a literal `(?<some>thing)` (by escaping it)?

> 2. Named groups are not mixed with unnamed groups.

Yes, I think it makes sense to either use named or unnamed groups, not both.

Reply:
> Not sure what you mean here. Do you mean it won't work to match a literal (?<some>thing) (by escaping it)?

Yes. If the pattern contains a character sequence that matches `(?<.+?>` but isn't actually a named group (e.g., `\\(?<name>` - parenthesis is escaped - or `\Q(?<.+?>\E` - pattern is quoted), then it won't work as intended.

> Yes, I think it makes sense to either use named or unnamed groups, not both.

There's no meaningful output to generate anyway (and this wasn't supported before either), but now it would also lead to wrong output as the group counts may be off. E.g., `(?<a>.)(.)(?<c>.)` with input `abc` produces `{"a":"a","c":"b"}`, while previously it was `{"a":"a","c":"c"}`

Reply:
Right, yes I think these are both reasonable, +1 for https://github.com/metafacture/metafacture-fix/commit/5395e907ee69a9d61f3a5dcb705d30cfa7b2e840, thanks for working on that!

Reply:
> as the group counts may be off.

Wait a minute, why do we even use the group index when we've got the name! Fixed in 26d2a9e (plus tests for these edge cases).